File: barnsley_fractal.py
import turtle as fern
import random
import sys
import json

try: # This adds ability to visualize the fern without having to install pillow in case the user don't want the
    # functionality to export the result.
    from PIL import ImageGrab
except ImportError:
    ImageGrab = None

default_preferences = {
    "ultrafast": True,
    "x_offset": 0,
    "y_offset": -230,
    "scale": 50,
    "plot_points": 10000,
    "probabilities": [0.01, 0.85, 0.07, 0.07],
    "x": 0,
    "y": 0,
    "speed": 0,
    "leftleaf_color": "#6E4C21",
    "rightleaf_color": "#37B469",
    "base_color": "#988a4f",
    "top_color": "#788511",
    "completion_message": None,
    "background_color": "black",
    "save_image": False
}


# Definition of the function to get the coordinates of the window
def window_coordinates():
    """ This function is solely to get the display size without adding any additional dependencies. What is does is it takes a screenshot of the screen and gets the size of the screenshot using python pillows utilities."""
    if ImageGrab:
        display_width, display_height = ImageGrab.grab().size
        left = (display_width - fern.window_width()) // 2
        upper = (display_height - fern.window_height()) // 2
        right = left + fern.window_width()
        lower = upper + fern.window_height()
        return (left, upper, right, lower)


# Definition of the Function to Check  command line arguments
def check_cmd_args(sys_args, arg_count):
    global preferences
    if arg_count > 1 and sys_args[1] == "-s":
        preferences["ultrafast"] = False
        print("plotting in slow mode.")
        if arg_count == 3:
            preferences["speed"] = int(sys_args[2])

    elif arg_count == 2 and sys_args[1] == "-u":
        preferences["ultrafast"] = True
        print("Plotting in ultra fast mode.")

    elif arg_count == 2 and sys_args[1] == "save":
        preferences["save_image"] = True

    elif arg_count == 3 and sys_args[1] == "-p" and sys_args[2].isdigit():
        preferences["plot_points"] = int(sys_args[2])
        print("plotting", sys_args[2], "points.")

    elif len(sys_args) == 2 and sys_args[1] == "-h":
        print("Refer the README.md file for more information.")
        sys.exit()

    elif arg_count > 2 and sys_args[1] == "-c":
        for arg in sys_args[2:]:
            if len(arg) == 9 and arg[2] == "#" and (int(arg[3:], 16) < 16777216):
                if arg[:2].lower() == "ll":
                    preferences["leftleaf_color"] = arg[2:]
                elif arg[:2].lower() == "rl":
                    preferences["rightleaf_color"] = arg[2:]
                elif arg[:2].lower() == "bs":
                    preferences["base_color"] = arg[2:]
                elif arg[:2].lower() == "tp":
                    preferences["top_color"] = arg[2:]

    elif arg_count == 2 and sys_args[1] == "reset":
        preferences = default_preferences
        print("Preferences set to default.")
    # Definitions of the functions for the fractal transformations.


def function1(x, y):
    fern.pencolor(preferences["base_color"])
    x1 = 0 * x + 0 * y
    y1 = 0 * x + 0.16 * y
    return x1, y1


def function2(x, y):
    fern.pencolor(preferences["top_color"])
    x2 = 0.85 * x + 0.04 * y + 0
    y2 = (-0.04) * x + 0.85 * y + 1.6
    return x2, y2


def function3(x, y):
    fern.pencolor(preferences["leftleaf_color"])
    x3 = 0.2 * x + (-0.26) * y + 0
    y3 = 0.23 * x + 0.22 * y + 1.6
    return x3, y3


def function4(x, y):
    fern.pencolor(preferences["rightleaf_color"])
    x4 = (-0.15) * x + 0.28 * y + 0
    y4 = 0.26 * x + 0.24 * y + 0.44
    return x4, y4


# end of function definitions


# Load preferences from the JSON file if there is any.
try:
    with open("preferences_barnsley.json", "r") as file:
        preferences = json.load(file)
except FileNotFoundError:
    print("Using default preferences.")
    preferences = default_preferences
except Exception as e:
    print(f"Error: {e}")
    sys.exit(1)
# end of loading preferences

# Changing preferences according to the command line arguments.
arg_count = len(sys.argv)
check_cmd_args(sys.argv, arg_count)

x = preferences["x"]
y = preferences["y"]
functions = [function1, function2, function3, function4]

# Plotting related code starts here.
fern.bgcolor(preferences["background_color"])
fern.speed(preferences["speed"])
fern.hideturtle()
fern.title("Barnsley Fern")

if preferences.get("ultrafast", True):
    fern.tracer(0)
try:
    for step in range(preferences["plot_points"]):
        fern.penup()
        fern.goto(x * preferences["scale"] + preferences["x_offset"], y * preferences["scale"] + preferences[
            "y_offset"])  # plotting the point after calculations with scale and offset
        fern.pendown()
        fern.circle(1)
        next_function = random.choices(functions, preferences["probabilities"])[
            0]  # randomly choosing the next function according to the probabilities.
        x, y = next_function(x, y)
        fern.update()
    if preferences["completion_message"]:
        fern.penup()
        fern.home()
        fern.goto(0, preferences["y_offset"] - 50)
        fern.pendown()
        fern.write(preferences["completion_message"], font=("Arial", 16, "normal"), align="center")

    if ImageGrab and preferences["save_image"]:
        # Capturing and saving the captured image as a PNG file
        image = ImageGrab.grab(bbox=window_coordinates())
        image.save("image_barnsley.png", "PNG")
    fern.exitonclick()

except fern.Terminator:
    print("Visualisation window closed by the user.")
# end of plotting code
finally:
    # Save preferences to a JSON file
    if arg_count > 1:
        try:
            with open("preferences_barnsley.json", "w") as file:
                json.dump(preferences, file)
            print("Your preferences have been saved.")
        except:
            print("Your preferences did not save.")

File: barnsley_fractal.jl
using JSON
using Makie
using StatsBase
using GLMakie


# print(ARGS)
default_preferences = Dict(
    "ultrafast" => true,
    "x_offset" => 0,
    "y_offset" => -230,
    "scale" => 50,
    "plot_points" => 10000,
    "probabilities" => [0.01, 0.85, 0.07, 0.07],
    "x" => 0,
    "y" => 0,
    "speed" => 0,
    "leftleaf_color" => "#6E4C21",
    "rightleaf_color" => "#37B469",
    "base_color" => "#988a4f",
    "top_color" => "#788511",
    "completion_message" => nothing,  # Julia's equivalent to None
    "background_color" => "black",
    "save_image" => false,
    "point_size" => 1,
    "use_dot_points" => false
)
  

# #Definition of the function to check the command line arguments.
function check_cmd_args(sys_args, arg_count)
    global preferences
    if arg_count > 0 && sys_args[1] == "-s"
        preferences["ultrafast"] = false
        print("plotting in slow mode.")
        if arg_count == 2
            preferences["speed"] = parse(Int, sys_args[2])
        end

    elseif arg_count == 1 && sys_args[1] == "-u"
        preferences["ultrafast"] = true
        print("plotting in ultrafast mode.")

    elseif arg_count == 1 && sys_args[1] == "save"
        preferences["save_image"] = true

    elseif arg_count == 2 && sys_args[1] == "-p" & args[2].isdigit()
        preferences["plot_points"] = parse(Int, sys_args[2])
        print("plotting ", sys_args[2], " points.")

    elseif arg_count == 1 &&sys_args[1] == "-h"
        print("Refer the README.md file for more information.")
        exit()

    elseif arg_count > 1 && sys_args[1] == "-c"
        for arg in sys_args[2:end]
            if length(arg) == 9 && arg[3] == "#" && (parse(Int, arg[4:end], base=16) < 16777216)
                if arg[1:2] == "ll"
                    preferences["leftleaf_color"] = arg[4:end]
                elseif arg[1:2] == "rl"
                    preferences["rightleaf_color"] = arg[4:end]
                elseif arg[1:2] == "bs"
                    preferences["base_color"] = arg[4:end]
                elseif arg[1:2] == "tp"
                    preferences["top_color"] = arg[4:end]
                end
            end
        end
    elseif arg_count == 1 && sys_args[1] == "reset"
        preferences = default_preferences
        print("Preferences reset to default.")
    end
end

# Definitions of the functions for the fractal transformations.
function function1(x, y)
    return [0.0 0.0; 0.0 0.16] * [x; y]
end

function function2(x, y)
    return [0.85 0.04; -0.04 0.85] * [x; y] + [0.0; 1.6]
end

function function3(x, y)
    return [0.2 -0.26; 0.23 0.22] * [x; y] + [0.0; 1.6]
end

function function4(x, y)
    return [-0.15 0.28; 0.26 0.24] * [x; y] + [0.0; 0.44]
end
#End of the function definitions.

# Loading preferences from the preferences from the JSON file if there is any.
try
    global preferences = JSON.parsefile("preferences_barnsley.json")
    print("prefs: "*preferences)
catch e
    if isa(e, SystemError) & occursin("2, nothing", string(e))  # Julia's equivalent to FileNotFoundError
        println("The file 'preferences_barnsley.json' does not exist. Using default preferences.")
        global preferences = default_preferences
    else
        println("Error: ", e) 
        println("Stacktrace: ", catch_backtrace())
        exit()
    end
end
# End of loading preferences.

# Adjusting any parameters that were passed as command line arguments if needed.
# To be implemented

#Changing preferences according to the command line arguments.  
arg_count = length(ARGS)
check_cmd_args(ARGS, arg_count)

# Unlike inn python implementation. This implementation stores all the points.
x_points = []
y_points = []

push!(x_points,preferences["x"])
push!(y_points,preferences["y"])
functions = [function1, function2, function3, function4]

# Create a ProbabilityWeights object from the list of weights
weights = ProbabilityWeights(preferences["probabilities"])



for i in 1:preferences["plot_points"]
    # Sample a function from the list of functions, weighted by the probabilities in the ProbabilityWeights object
    selected_function = sample(functions, weights)

    # Execute the selected function
    println(x_points[end],y_points[end])
    x,y = selected_function(x_points[end],y_points[end])
    push!(x_points,x)
    push!(y_points,y)
end

plot = scatter(Float32.(x_points),Float32.(y_points))
display(plot)


if arg_count > 0
    try
        JSON.open("preferences_barnsley.json", "w") do file
            JSON.print(file, preferences)
        end
        println("Preferences saved to preferences_barnsley.json.")
    catch e
        println("Error saving preferences: ", e)
    end
end